crates/c-api/src/wasi.rs

```diff
@@ -270,3 +327,89 @@ pub unsafe extern "C" fn wasi_config_preopen_dir(
 
     true
 }
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_pipe_new(
+    limit: usize,
+    ret_read_pipe: *mut *mut wasi_read_pipe_t,
+    ret_write_pipe: *mut *mut wasi_write_pipe_t,
+) {
+    let queue = BoundedVecDeque::new(limit);
+    let queue = Arc::new(RwLock::new(queue));
+
+    if !ret_read_pipe.is_null() {
+        *ret_read_pipe = Box::into_raw(Box::new(wasi_read_pipe_t {
+            queue: queue.clone(),
+        }));
+    }
+    if !ret_write_pipe.is_null() {
+        *ret_write_pipe = Box::into_raw(Box::new(wasi_write_pipe_t {
+            queue: queue.clone(),
+        }));
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn wasi_read_pipe_len(read_pipe: &wasi_read_pipe_t) -> usize {
+    let queue = read_pipe.queue.read().unwrap();
+    queue.len()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_read_pipe_read(
+    read_pipe: &mut wasi_read_pipe_t,
+    buf: *mut u8,
+    buf_len: usize,
+) -> usize {
+    let mut buf = crate::slice_from_raw_parts_mut(buf, buf_len);
+    let mut queue = read_pipe.queue.write().unwrap();
+    std::io::Read::read(&mut *queue, &mut buf).unwrap()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_write_pipe_write(
+    write_pipe: &mut wasi_write_pipe_t,
+    buf: *const u8,
+    buf_len: usize,
+) -> usize {
+    let buf = crate::slice_from_raw_parts(buf, buf_len);
+    let mut queue = write_pipe.queue.write().unwrap();
+    std::io::Write::write(&mut *queue, buf).unwrap()
+}
+
+#[derive(Default)]
+pub struct BoundedVecDeque<T> {
+    deque: VecDeque<T>,
+    limit: usize,
+}
+impl<T> BoundedVecDeque<T> {
+    fn new(limit: usize) -> Self {
+        Self {
+            deque: VecDeque::new(),
+            limit,
+        }
+    }
+    #[inline]
+    fn len(&self) -> usize {
+        self.deque.len()
+    }
+}
+impl io::Read for BoundedVecDeque<u8> {
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        io::Read::read(&mut self.deque, buf)
+    }
+}
+impl io::Write for BoundedVecDeque<u8> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let amt = self.limit.saturating_sub(self.len());
+        let amt = std::cmp::min(amt, buf.len());
+        self.deque.extend(&buf[..amt]);
+        Ok(amt)
+    }
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        self.deque.flush()
```

Comment:
I think this can be omitted since it doesn't do anything

Reply:
`std::io::Write::flush` has not default body. IMO to force implementers to thinks about it.

Reply:
Yes that makes sense but `VecDeque::flush` is a noop.